dao/mongodb.rya/src/main/java/org/apache/rya/mongodb/iter/AggregationUtil.java

```diff
@@ -0,0 +1,322 @@
+
+        final BasicDBObject setIntersectionExists =
+            gt(
+                size(
+                    setIntersection(
+                        documentVisibilityField,
+                        authOrList.toArray()
+                    )
+                ),
+                0
+            );
+
+        final BasicDBObject orExpression = or(setIsSubset, setIntersectionExists);
+
+        final List<DBObject> pipeline = new ArrayList<>();
+        pipeline.add(
+            redact(
+               orExpression,
+               "$$DESCEND",
+               "$$PRUNE"
+            )
+        );
+
+        return pipeline;
+    }
+
+    /**
+     * Creates all combinations of the values that are of the size of value
+     * array or smaller without repetition.
+     * @param values the {@link List} of values to create combinations from.
+     * @return the {@link List} of combinations.
+     */
+    public static <T> List<List<T>> createCombinations(final List<T> values) {
+        final List<List<T>> allCombinations = new ArrayList<>();
+        for (int i = 1; i <= values.size(); i++) {
+            allCombinations.addAll(createCombinations(values, i));
+        }
+        return allCombinations;
+    }
+
+    /**
+     * Creates all combinations of the values that are of the specified size
+     * without repetition.
+     * @param values the {@link List} of values to create combinations from.
+     * @param size the size of the combinations.
+     * @return the {@link List} of combinations.
+     */
+    public static <T> List<List<T>> createCombinations(final List<T> values, final int size) {
+        if (0 == size) {
+            return Collections.singletonList(Collections.<T> emptyList());
+        }
+
+        if (values.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        final List<List<T>> combination = new LinkedList<List<T>>();
+
+        final T actual = values.iterator().next();
+
+        final List<T> subSet = new LinkedList<T>(values);
+        subSet.remove(actual);
+
+        final List<List<T>> subSetCombination = createCombinations(subSet, size - 1);
+
+        for (final List<T> set : subSetCombination) {
+            final List<T> newSet = new LinkedList<T>(set);
+            newSet.add(0, actual);
+            combination.add(newSet);
+        }
+
+        combination.addAll(createCombinations(subSet, size));
+
+        return combination;
+    }
+
+    /**
+     * Creates an "if-then-else" MongoDB expression.
+     * @param ifStatement the "if" statement {@link BasicDBObject}.
+     * @param thenResult the {@link Object} to return when the
+     * {@code ifStatement} is {@code true}.
+     * @param elseResult the {@link Object} to return when the
+     * {@code ifStatement} is {@code false}.
+     * @return the "if" expression {@link BasicDBObject}.
+     */
+    public static BasicDBObject ifThenElse(final BasicDBObject ifStatement, final Object thenResult, final Object elseResult) {
```

Comment:
These functions aren't available in BasicDBObjectBuilder.  I'll make the methods that aren't aggregation specific private.